describe REST API; fixes #490

LightIsLife opened the pull request:
describe the value of the REST API and showcase the possibilities; replacing entire current content;

Signed-off-by: Markus Lipp lipp.markus@gmail.com (github: LightIsLife)

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,29 @@ title: openHAB REST API
-It can be used to integrate openHAB with other systems as it allows read access to items and item states as well as status updates or the sending of commands for items.
-Furthermore, it gives access to the sitemaps, so that it is the interface to be used by remote user interfaces (e.g. fat clients or javascript based web clients).
-The REST API also supports server-push, so you can subscribe yourself on change notification for certain resources.
+The REST API gives the ability to access a wide range of information associated with a running openHAB installation, and the ability to invoke changes to the openHAB installation. 
```
"access a wide range of information associated with a running openHAB installation"

this seems very generic. How about a simple first sentence like (not polished):

> The openHAB [REST API](https://en.wikipedia.org/wiki/REST_API) offers access to most aspects of the openHAB system. Data related to, e.g., Items, Things and Bindings can be accessed and actions can be performed via a simple http based interaction method.

bgilmer77 replied:
Or, how about this...
The [REST](https://en.wikipedia.org/wiki/Representational_state_transfer) API provides both the ability to access a wide range of information associated with a running openHAB installation, and the ability to invoke changes to the openHAB installation.
This summarizes both the ability to access information and set information.  It also provides a link to a wikipedia article on REST, allowing us to cut out the definition of REST further down in the article.

LightIsLife replied:
sounds good, done; adding the link to wikipedia....but am I remembering correctly that links to sources outside the OH universe are discouraged? At least, I thought reading it...which is why I added some description in the first place. Linking it is more elegant though.....

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,29 @@ title: openHAB REST API
-Furthermore, it gives access to the sitemaps, so that it is the interface to be used by remote user interfaces (e.g. fat clients or javascript based web clients).
-The REST API also supports server-push, so you can subscribe yourself on change notification for certain resources.
+The REST API gives the ability to access a wide range of information associated with a running openHAB installation, and the ability to invoke changes to the openHAB installation. 
+The REST API provides an interface that may only be accessed programmatically.
```
may only?

LightIsLife replied:
may only ...as in allowed....will change to can

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,29 @@ title: openHAB REST API
-The REST API also supports server-push, so you can subscribe yourself on change notification for certain resources.
+The REST API gives the ability to access a wide range of information associated with a running openHAB installation, and the ability to invoke changes to the openHAB installation. 
+The REST API provides an interface that may only be accessed programmatically.
+The information available from the API includes Items and their State, installed bindings, sitemap information, and much more. 
```
state lowercase

bgilmer77 replied:
So wrapping the information in these lines along with the important information about server-push, how about:

The information available from the API includes Items and their state, installed bindings, sitemap information, and much more. The information exchange takes place through HTTP and the API provides a standardized platform for information exchange among different computer systems.  The REST API also supports server-push, so you can subscribe to certain resources and receive change notification for those resources when they occur.

LightIsLife replied:
changed

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,29 @@ title: openHAB REST API
+After the REST API documentation add-on is installed, it is accessible through the openHAB dashboard. 
+Access the REST API by connecting any browser to: http://123.456.789.000/8080/doc/index.html (with 123.456.789.000 being the ip address of the computer that runs openHAB). 
+Access over the Internet to the REST API is possible but this represents a significant security risk, and users are encouraged to ensure safe and secure connections. 
+Some ways this can be achieved include using [myopenHAB cloud services](http://docs.openhab.org/installation/security.html#myopenhab-cloud-service), a [reverse proxy server](http://docs.openhab.org/installation/security.html#nginx-reverse-proxy), or a VPN.
```
The whole article would look better with a few headlines.

ThomDietrich replied:
As I said before, I don't think the way how a user accesses the REST API belongs here. We would need to add similar content to the Basic UI, Paper UI, Classic UI, Android app, iOS app, ... articles. There is already one consolidated article discussing the whole issue. You can simply link it here 😃  http://docs.openhab.org/installation/security.html

bgilmer77 replied:
+1  re: Headlines, perhaps it is best to get the content down and then add the headlines?

+1 deleting the information about accessing and just pointing to the security article.

LightIsLife replied:
done, removed the individual examples; added headlines, but it is a short article not sure how many you had in mind

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,29 @@ title: openHAB REST API
+- Updating Item States through external devices that are capable of sending http commands (for example, certain motion detectors or surveillance cameras)
 
-{% include contribution-wanted.html %}
+The REST API is very intuitives and excellently documented and it would vastly exceed this space to present all possible example, nevertheless, here an example to switch an item of the Switch Type (such as a lamp) on and off.
```
careful, it is the "openHAB REST API Documentation **frontend**" you are talking about here

ThomDietrich commented on configuration/restdocs.md:
````diff
@@ -7,14 +7,29 @@ title: openHAB REST API
+- Switching ```My_Item``` OFF: ```curl -X POST --header "Content-Type: text/plain" --header "Accept: application/json" -d "OFF" "http://123.456.789.000:8080/rest/items/My_Item"``` 
+- Switching ```My_Item``` ON: ```curl -X POST --header "Content-Type: text/plain" --header "Accept: application/json" -d "ON" "http://123.456.789.000:8080/rest/items/My_Item"``` 
+
+with ```123.456.789.000``` the ip address of the computer that runs openHAB.  
````
It's more common to use a standard IP or a hostname. I'd go with hostname, something like "openhab-device" or "openhab-host".

LightIsLife replied:
Done

ThomDietrich commented on configuration/restdocs.md:
````diff
@@ -7,14 +7,29 @@ title: openHAB REST API
-{% include contribution-wanted.html %}
+The REST API is very intuitives and excellently documented and it would vastly exceed this space to present all possible example, nevertheless, here an example to switch an item of the Switch Type (such as a lamp) on and off.
+- Switching ```My_Item``` OFF: ```curl -X POST --header "Content-Type: text/plain" --header "Accept: application/json" -d "OFF" "http://123.456.789.000:8080/rest/items/My_Item"``` 
+- Switching ```My_Item``` ON: ```curl -X POST --header "Content-Type: text/plain" --header "Accept: application/json" -d "ON" "http://123.456.789.000:8080/rest/items/My_Item"``` 
````
You should break these examples to multi-line code fences to gain colored highlighting :tada:

LightIsLife replied:
Done

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,29 @@ title: openHAB REST API
-The REST API of openHAB serves different purposes.
-It can be used to integrate openHAB with other systems as it allows read access to items and item states as well as status updates or the sending of commands for items.
-Furthermore, it gives access to the sitemaps, so that it is the interface to be used by remote user interfaces (e.g. fat clients or javascript based web clients).
-The REST API also supports server-push, so you can subscribe yourself on change notification for certain resources.
```
The last line here describes an important feature and should not be deleted, in fact I'd give it its own headline.

bgilmer77 replied:
I agree - and I included it in the version I posted in the discussion we had been having.

LightIsLife replied:
Thanks added it back in.

bgilmer77 replied:
Don't know how to highlight specific sections from below.  So here we go.  Below, recommend changing: are based on a simple http protocoll.

to this:
are based on http protocol.

bgilmer77 replied:
Ahh - now I found it - on my mac, press the blue '+' sign.  Geez - so much to learn.

LightIsLife commented:
@bgilmer77 sorry, looks like we had some overlap with you commenting on the review and me addressing the comments there (just had some time to do that); let me know if in my latest commit #2f137b0 there are still some outstanding issued

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
-Furthermore, it gives access to the sitemaps, so that it is the interface to be used by remote user interfaces (e.g. fat clients or javascript based web clients).
-The REST API also supports server-push, so you can subscribe yourself on change notification for certain resources.
+Through the openHAB [REST API](https://en.wikipedia.org/wiki/REST_API) most aspects of the openHAB system can be accessed by other programs.  
+This includes for example the access to all data related to Items, Things and Bindings as well as the capabilities to invoke  actions that change the state of Items. Interactions with the REST API are based on a simple http protocoll.
```
So - just practicing - change protocoll to protocol

bgilmer77 replied:
@LightIsLife it looks to me as if there are still a number of my comments which have not been addressed in your latest commit.  I hope I am looking at things correctly.  An example is that the word 'protocol' is still spelled 'protocoll' in the version I see when I use the 'view' button to see the entire document without all the diff coding.

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
-The REST API also supports server-push, so you can subscribe yourself on change notification for certain resources.
+Through the openHAB [REST API](https://en.wikipedia.org/wiki/REST_API) most aspects of the openHAB system can be accessed by other programs.  
+This includes for example the access to all data related to Items, Things and Bindings as well as the capabilities to invoke  actions that change the state of Items. Interactions with the REST API are based on a simple http protocoll.
+The REST API provides an interface that can only be accessed programmatically.
```
only?  Is this correct?  @ThomDietrich

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+This includes for example the access to all data related to Items, Things and Bindings as well as the capabilities to invoke  actions that change the state of Items. Interactions with the REST API are based on a simple http protocoll.
+The REST API provides an interface that can only be accessed programmatically.
+The information available from the API includes Items and their state, installed bindings, sitemap information, and much more. 
+The information exchange takes place through HTTP and the API provides a defined protocoll for information exchange among different computer systems. 
```
This information is repetitive.  I know I reformatted from the beginning to this point to address this.  @LightIsLife , you might want to look at what I did to see if you want to adopt more of that wording to reduce repeated information such as referring again to HTTP.

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+The information available from the API includes Items and their state, installed bindings, sitemap information, and much more. 
+The information exchange takes place through HTTP and the API provides a defined protocoll for information exchange among different computer systems. 
+However, from a technical perspective it is very easy to use. 
+Be aware that the REST API is not automatically installed in openHAB. 
```
I think in my version I briefly mention how to install it.  This is important as someone may be a new user and have come directly here to start working.  So a brief mention of where it is located would be good.  I think the wording I used after @ThomDietrich 's comment was:

Be aware that the REST API is not automatically installed in openHAB. After the REST API documentation add-on is installed, it is accessible through the openHAB dashboard.

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+The information exchange takes place through HTTP and the API provides a defined protocoll for information exchange among different computer systems. 
+However, from a technical perspective it is very easy to use. 
+Be aware that the REST API is not automatically installed in openHAB. 
+Access over the Internet to the REST API is possible but this represents a significant security risk, and users are encouraged to [ensure safe and secure connections](http://docs.openhab.org/installation/security.html). 
```
Very nice change - clean.

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+Be aware that the REST API is not automatically installed in openHAB. 
+Access over the Internet to the REST API is possible but this represents a significant security risk, and users are encouraged to [ensure safe and secure connections](http://docs.openhab.org/installation/security.html). 
+
+## Examples for the Use of the REST API
```
How about, "REST API Examples"?

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+Access over the Internet to the REST API is possible but this represents a significant security risk, and users are encouraged to [ensure safe and secure connections](http://docs.openhab.org/installation/security.html). 
+
+## Examples for the Use of the REST API
+Some tasks that can be accomplished by using the REST API include:
```
Maybe with adding a header, you can now eliminate this sentence?  It seems to be a repetition.

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+Some tasks that can be accomplished by using the REST API include:
+
+- Using third party software on cell phones, such as tasker to [open your garage door](https://community.openhab.org/t/triggering-items-using-openhab-2s-rest-api-from-tasker/14027)
+- Interacting with openHAB via other programs; several computer languages offer libraries to easily access REST APIs generally, including openHAB’s REST API
```
Yuck - I wrote it, I don't like it.  Consider revising "several computer languages offer libraries to easily access REST APIs generally, including openHAB's REST API" to:

many computer languages can easily make use of REST APIs

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+- Using third party software on cell phones, such as tasker to [open your garage door](https://community.openhab.org/t/triggering-items-using-openhab-2s-rest-api-from-tasker/14027)
+- Interacting with openHAB via other programs; several computer languages offer libraries to easily access REST APIs generally, including openHAB’s REST API
+- Accessing of information from devices that support http communication
+- Updating Item States through external devices that are capable of sending http commands (for example, certain motion detectors or surveillance cameras)
```
Consider "(for example, some motion detectors...

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+- Accessing of information from devices that support http communication
+- Updating Item States through external devices that are capable of sending http commands (for example, certain motion detectors or surveillance cameras)
+
+The openHAB REST API frontend is very intuitives and excellently documented and it would vastly exceed this space to present all possible example, nevertheless, here an example to switch an item of the Switch Type (such as a lamp) on and off.
```
Consider, front end is very intuitive and...

Consider, all possible examples...

Consider, here is an example that switches an Item of the Switch Type...

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,38 @@ title: openHAB REST API
+One final remark - when openHAB’s REST API is accessed from another domain, the following error may be encountered: [No 'Access-Control-Allow-Origin' header is present on the requested resource](https://community.openhab.org/t/cors-problem-at-rest-api/3712/10). 
+In this case, it may be necessary to add the line org.eclipse.smarthome.cors:enable=true in the file services/runtime.cfg.
+
+The REST API also supports server-push, so you can subscribe yourself on change notification for certain resources.
```
I think it is better to put the server-push capability up at the beginning - please consider this.

bgilmer77 commented on configuration/restdocs.md:
````diff
@@ -7,14 +7,38 @@ title: openHAB REST API
 
-{% include contribution-wanted.html %}
+```java
+curl -X POST --header "Content-Type: text/plain" --header "Accept: application/json" -d "ON" "http://ip_openHAB_host:8080/rest/items/My_Item"
````
Once we get the basics of the article together, I would like to include a reference to the API documentation and specifically how to access the curl and HTTP examples you can find using the "Try It" button.  But this can wait.

LightIsLife commented:
Got it, will work through your comments hopefully tonight. Thanks!!!

On Wed, Sep 20, 2017, 10:59 bgilmer77 <notifications@github.com> wrote:

> *@bgilmer77* commented on this pull request.
> ------------------------------
>
> In configuration/restdocs.md
> <https://github.com/openhab/openhab-docs/pull/508#discussion_r139912525>:
>
> > @@ -7,14 +7,38 @@ title: openHAB REST API
>
>  # openHAB REST API
>
> -The REST API of openHAB serves different purposes.
> -It can be used to integrate openHAB with other systems as it allows read access to items and item states as well as status updates or the sending of commands for items.
> -Furthermore, it gives access to the sitemaps, so that it is the interface to be used by remote user interfaces (e.g. fat clients or javascript based web clients).
> -The REST API also supports server-push, so you can subscribe yourself on change notification for certain resources.
> +Through the openHAB [REST API](https://en.wikipedia.org/wiki/REST_API) most aspects of the openHAB system can be accessed by other programs.
> +This includes for example the access to all data related to Items, Things and Bindings as well as the capabilities to invoke  actions that change the state of Items. Interactions with the REST API are based on a simple http protocoll.
>
> @LightIsLife <https://github.com/lightislife> it looks to me as if there
> are still a number of my comments which have not been addressed in your
> latest commit. I hope I am looking at things correctly. An example is that
> the word 'protocol' is still spelled 'protocoll' in the version I see when
> I use the 'view' button to see the entire document without all the diff
> coding.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab-docs/pull/508#discussion_r139912525>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AXtFf0NWwgEG0fy5yHyIB1uyd8y9Zofnks5skNQEgaJpZM4Pcu8I>
> .
>

LightIsLife commented:
@bgilmer77 hope I addressed your concerns in commit 671e6a7
Thanks for your review

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,45 @@ title: openHAB REST API
+Be aware that the REST API is not automatically installed in openHAB. 
+After installation, the REST APT is accessible through the openHAB dashboard.
+Access over the Internet to the REST API is possible, but this represents a significant security risk. 
+Users are encouraged to [ensure safe and secure connections](http://docs.openhab.org/installation/security.html). 
```
Nice ;)

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,45 @@ title: openHAB REST API
+- Using third party software on cell phones, such as tasker to [open your garage door](https://community.openhab.org/t/triggering-items-using-openhab-2s-rest-api-from-tasker/14027)
+- Interacting with openHAB via other programs; many computer languages can easily make use of REST APIs
+- Accessing of information from devices that support http communication
+- Updating Item states through external devices that are capable of sending http commands (for example, some motion detectors or surveillance cameras)
```
I'd start with the generic examples first and move the cell phone example to the bottom.

As for the generic ones, how about something like:

*    Retrieve openHAB data from external applications
*    Inject data and trigger events in openHAB from external applications
*    Inspect openHAB Bindings/Things or Items, learn about current states, parameters or problems

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,45 @@ title: openHAB REST API
-- The interactive REST API documentation available through the openHAB 2 dashboard (Add-on "misc-restdocs", installed by default)
-- [The openHAB 1.x wiki REST API article](https://github.com/openhab/openhab/wiki/REST-API)
+- Using third party software on cell phones, such as tasker to [open your garage door](https://community.openhab.org/t/triggering-items-using-openhab-2s-rest-api-from-tasker/14027)
+- Interacting with openHAB via other programs; many computer languages can easily make use of REST APIs
```
via -> from
programming languages and automation tools
remove easily

bgilmer77 commented:
@LightIsLife I have no objection to posting this as-is.  It is much better than what was there before.  I have some ideas about how to improve it, but these are just layout and minor issues.  So I would go ahead with what you have at this point.
One comment - I would provide an embedded link to the REST documentation you refer to at the end of the article, if that is not too much trouble.

ThomDietrich commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,45 @@ title: openHAB REST API
+- Updating Item states through external devices that are capable of sending http commands (for example, some motion detectors or surveillance cameras)
 
-{% include contribution-wanted.html %}
+The openHAB REST API frontend is very intuitive and excellently documented.
```
The "frontend" was not clearly addressed yet. Could you add an explanation to the introduction stating, that openHAB provides a REST API and a REST API documentation and interaction web frontend?

bgilmer77 replied:
Yes - perhaps just delete the word frontend?

ThomDietrich commented on configuration/restdocs.md:
````diff
@@ -7,14 +7,45 @@ title: openHAB REST API
+
+```java
+curl -X POST --header "Content-Type: text/plain" --header "Accept: application/json" -d "ON" "http://ip_openHAB_host:8080/rest/items/My_Item"
+``` 
````
Instead of showing these two redundant examples, how about a second GET example showing the new state?

ThomDietrich replied:
Also: You are showing curl examples without any explanation. It would probably to shortly say that you are issuing an http [POST]() request using [curl](), the famous command line tool available for many platforms.

bgilmer77 commented on configuration/restdocs.md:
```diff
@@ -7,14 +7,45 @@ title: openHAB REST API
+This includes for example, the access to all data related to Items, Things and Bindings as well as the capabilities to invoke  actions that change the state of Items. 
+Interactions with the REST API are based on the http protocol and therefore can exchange among different computer systems. 
+The information available from the API includes Items and their state, installed bindings, sitemap information, and much more. 
+The REST API provides an interface that can only be accessed programmatically.
```
+1

LightIsLife commented:
@ThomDietrich @bgilmer77 thanks for your comments
I hope I addressed all your comments; not so sure about a screenshot, seems pretty redundant with the examples. But more to the point, won't have the time to do that for the next few days as I will be travelling

ThomDietrich commented:
The last commit makes everything round and clear. Great.
Okay, let's not torture you any longer... Let's merge and wish you a great trip! 😉
